File: GO/a2zSheet/step1_basics/recurrsion.go
package main

import (
	"bufio"
	. "fmt"
	"os"
)

var in = bufio.NewReader(os.Stdin)
var out = bufio.NewWriter(os.Stdout)

func PrintName(n int) {
	if n == 0 {
		return
	}
	Fprint(out, "Bhushan\n")
	PrintName(n - 1)
}

func PrintNums_12N(n int) {
	if n == 0 {
		return
	}
	PrintNums_12N(n - 1)
	Print(n, " ")
}

func PrintNums_N21(n int) {
	if n == 0 {
		return
	}
	Fprintln(out, n)
	PrintNums_N21(n - 1)
}

func Sum(n int) int {
	if n == 0 {
		return 0
	}

	return n + Sum(n-1)
}

func Fact(n int) int {
	if n == 0 {
		return 1
	}

	return n * Fact(n-1)
}

func ReverseArr(arr []int, start int, end int) {
	if start >= end {
		return
	}
	var temp = arr[start]
	arr[start] = arr[end]
	arr[end] = temp
	ReverseArr(arr, start+1, end-1)
}

func IsPal(s string, start int, end int) bool {
	if start >= end {
		return true
	}

	return (s[start] == s[end] && IsPal(s, start+1, end-1))
}

func Fib(n int, a int, b int, counter int) {
	if n == counter {
		return
	}
	c := a + b
	a = b
	b = c
	Fprint(out, c, " ")
	Fib(n, a, b, counter+1)
}

func solve() {
	var n int
	Fscan(in, &n)
	// arr := make([]int, n)
	// for i := 0; i < n; i++ {
	// 	Fscan(in, &arr[i])
	// }
	// var s string
	// Fscan(in, &s)
	Fprint(out, "0 1 ")
	Fib(n, 0, 1, 1)

}

func main() {
	defer out.Flush()
	var t int

	for Fscan(in, &t); t > 0; t-- {
		solve()
	}
}

File: a2zSheet/step1_basics/CPP/recurrsion.cpp
#include<bits/stdc++.h>

using namespace std;

void printName(int n){
    if(n==0)return;
    cout<<"Bhushan\n";
    printName(n-1);
}

void printNums_12N(int n){
    if(n==0)return;
    printNums_12N(n-1);
    cout<<n<<" ";
}

void printNums_N21(int n){
    if(n==0)return;
    cout<<n<<" ";
    printNums_N21(n-1);
}

int sum(int n){
    if(n==0)return 0;
    return n + sum(n-1);
}

int fact(int n){
    if(n==0)return 1;
    return n*fact(n-1);
}

void reverseArray(vector<int>&v,int start,int end){
    if(start>=end)return;
    swap(v[end],v[start]);
    reverseArray(v,start+1,end-1);
}

bool isPal(string &s, int start,int end){
    if(start>=end)return true;
    return s[start]==s[end] && isPal(s,start+1,end-1);
}

void fib(int a,int b,int n){
    if(n==0)return;
    int c = a+b;
    a = b;
    b = c;
    cout<<c<<" ";
    fib(a,b,n-1);

}


int main(){
    int n;cin>>n;
    // vector<int>v(n);
    // for(auto &c:v)cin>>c;
    // reverseArray(v,0,n-1);
    // for(auto &c:v)cout<<c<<" ";
    cout<<"0 1 ";
    fib(0,1,n-2);
    return 0;
}